### ISSUE TITLE
Auth: Add support for multiple auth mechanisms

### ISSUE THREAD

**Issue by demianbrecht** [simple-salesforce/simple-salesforce#114]
Currently we're only supporting SOAP-based auth. It would be handy to support both SOAP and OAuth 2.0 (at least). Rough thoughts on how to implement this:

Create an `Authenticator` base class that each implementation (`soap.Authenticator`, `oauth.PasswordAuthenticator`, `oauth.AuthorizationAuthenticator`, etc) can inherit from. Implement a common `Tokens` structure that contains access and refresh tokens. Implement `authenticate` and `authorize` on the authenticators as required. Have them return the populated `Tokens` struct. Access token can be used in requests, refresh token can be used to auto-refresh access token as needed.

**Comment by nickcatal** [comment]
Love the idea of implementing OAuth2.

Do you think the way the OAuth flow works maps well enough to the way the Soap Username/Password/Token flow works to have them use the same base class? There are no refresh tokens with the SOAP, for example.

By the way, here's some (untested) code I wrote up a long time ago that handles oauth2. https://github.com/neworganizing/salesforce-oauth2/blob/master/salesforce_oauth2/auth.py

**Comment by demianbrecht** [comment]
Yeah I do. Sadly, I'd fully implemented this but boneheadedly `rm -rf`d the directory because it was external to simple-salesforce. Both OAuth (all server-side flows) and SOAP must implement `.authenticate()`. They both return a token struct. In the case of SOAP, it returns an access token and `None` for the refresh token.

**Comment by nickcatal** [comment]
This makes sense to me. +1 to the idea

**Comment by pitbulk** [comment]
Here are more examples of the Oauth2 authentication:
https://github.com/django-salesforce/django-salesforce/blob/master/salesforce/auth.py
https://github.com/pitbulk/salesforce-python-sdk/blob/master/salesforce/login.py#L35
